### PR TITLE
[Merged by Bors] - feat (NumberTheory/LSeries): Dirichlet series as Mellin transforms 

### PR DESCRIPTION
**Pull request by loefflerd**
Show that the Mellin transform of a power series in `exp (-t)` is the corresponding Dirchlet series.

---

- [x] depends on: #11069
- [x] depends on: #11150 
[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    (h_sum : Summable fun i ↦ ‖a i‖ / (p i) ^ s.re) :
+    HasSum (fun i ↦ π ^ (-s) * Gamma s * a i / p i ^ s) (mellin F s) := by
+  have hs' : s ≠ 0 := fun h ↦ lt_irrefl _ (zero_re ▸ h ▸ hs)
+  let a' i := if p i = 0 then 0 else a i
+  have hp' i : a' i = 0 ∨ 0 < p i := by
+    simp only [a']
+    split_ifs with h <;> tauto
+    exact Or.inr (lt_of_le_of_ne (hp i) (Ne.symm h))
+  have (i t) : (if p i = 0 then 0 else a i * rexp (-π * p i * t)) =
+    (a' i * rexp (-π * p i * t)) := by simp only [a', ite_mul, zero_mul]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean:99]:
```suggestion
  have (i t) : (if p i = 0 then 0 else a i * rexp (-π * p i * t)) =
      (a' i * rexp (-π * p i * t)) := by
    simp only [a', ite_mul, zero_mul]
```
I think it is better to have the proof on the next line when the statement spans more than one line.

File: Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    · have := hp i
+      rw [norm_of_nonneg (by positivity)]
+
+/-- Tailored version for even Jacobi theta functions. -/
+lemma hasSum_mellin_pi_mul_sq {a : ι → ℂ} {r : ι → ℝ} {F : ℝ → ℂ} {s : ℂ} (hs : 0 < s.re)
+    (hF : ∀ t ∈ Ioi 0, HasSum (fun i ↦ if r i = 0 then 0 else a i * rexp (-π * r i ^ 2 * t)) (F t))
+    (h_sum : Summable fun i ↦ ‖a i‖ / |r i| ^ s.re) :
+    HasSum (fun i ↦ Gammaℝ s * a i / |r i| ^ s) (mellin F (s / 2)) := by
+  have hs' : 0 < (s / 2).re := by rw [div_ofNat_re]; positivity
+  have h (i) : r i ^ 2 = 0 ↔ r i = 0 := by simp
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean:118]:
```suggestion
  have h i : r i ^ 2 = 0 ↔ r i = 0 := by simp
```

File: Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  simp_rw [← h] at hF
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean:119]:
```suggestion
  have h i : r i = 0 ↔ r i ^ 2 = 0 := by simp
  simp_rw [h] at hF
```

File: Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+lemma hasSum_mellin_pi_mul_sq' {a : ι → ℂ} {r : ι → ℝ} {F : ℝ → ℂ} {s : ℂ} (hs : 0 < s.re)
+    (hF : ∀ t ∈ Ioi 0, HasSum (fun i ↦ a i * r i * rexp (-π * r i ^ 2 * t)) (F t))
+    (h_sum : Summable fun i ↦ ‖a i‖ / |r i| ^ s.re) :
+    HasSum (fun i ↦ Gammaℝ (s + 1) * a i * SignType.sign (r i) / |r i| ^ s)
+    (mellin F ((s + 1) / 2)) := by
+  have hs₁ : s ≠ 0 := fun h ↦ lt_irrefl _ (zero_re ▸ h ▸ hs)
+  have hs₂ : 0 < (s + 1).re := by rw [add_re, one_re]; positivity
+  have hs₃ : s + 1 ≠ 0 := fun h ↦ lt_irrefl _ (zero_re ▸ h ▸ hs₂)
+  have (i t) : (a i * r i * rexp (-π * r i ^ 2 * t)) = if r i = 0 then 0 else
+    (a i * r i * rexp (-π * r i ^ 2 * t)) := by split_ifs with h <;> simp [h]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean:140]:
```suggestion
  have (i t) : (a i * r i * rexp (-π * r i ^ 2 * t)) =
      if r i = 0 then 0 else (a i * r i * rexp (-π * r i ^ 2 * t)) := by
    split_ifs with h <;> simp [h]
```

**Comment by MichaelStollBayreuth** [comment]
There are a lot of fiddly things involving complex and/or real powers, norms, gamma functions, ...
I was wondering whether one could perhaps extract some more general lemmas to help with that. But maybe all these instances are sufficiently different from one another to really make this possible.

One thing to be added could be `Complex.ne_zero_of_zero_lt_re` (compare [Complex.ne_zero_of_one_lt_re](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Complex/Abs.html#Complex.ne_zero_of_one_lt_re)).

File: Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+import Mathlib.Analysis.PSeries
+/-!
+# Dirichlet series as Mellin transforms
+
+Here we prove general results of the form "the Mellin transform of a power series in exp (-t) is
+a Dirichlet series".
+-/
+
+open Filter Topology Asymptotics Real Set MeasureTheory
+open Complex hiding abs_of_nonneg
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean:17]:
```suggestion
open Complex
```
Hiding `abs_of_nonneg` does not seem to be necessary.

**Comment by MichaelStollBayreuth** [comment]
See, e.g., the private lemma `foo` [here](https://github.com/MichaelStollBayreuth/EulerProducts/blob/main/EulerProducts/LSeriesUnique.lean#L43) ...

File: Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean (hunk 3 above)
Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/MellinEqDirichlet.lean:119]: This line is now golfed away entirely

**Comment by loefflerd** [comment]
> One thing to be added could be `Complex.ne_zero_of_zero_lt_re` (compare [Complex.ne_zero_of_one_lt_re](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Complex/Abs.html#Complex.ne_zero_of_one_lt_re).

I'm not keen on this, the statement only gets used twice (+ once more with a slight variant) and the proof term is less than half a line long. I don't think it's worthy of being a lemma of its own. Anyway, where would we stop? If we have `Complex.ne_zero_of_zero_lt_re` and `Complex.ne_zero_of_one_lt_re`, are we going to have `Complex.ne_zero_of_seventeen_point_six_lt_re` as well?

**Comment by MichaelStollBayreuth** [comment]
The conditions `0 < s.re` and `1 < s.re` will be far more frequent than any other such bound, I would think, so there should not be much danger of this getting out of hand. But if you think you won't need it anywhere else, it is of course OK to leave it inlined.

**Comment by loefflerd** [comment]
> The conditions `0 < s.re` and `1 < s.re` will be far more frequent than any other such bound, I would think, so there should not be much danger of this getting out of hand. But if you think you won't need it anywhere else, it is of course OK to leave it inlined.

All of these non-vanishing statements are needed purely as inputs to `zero_cpow`, to check that `0 ^ s` terms don't cause any trouble. They could be got rid of entirely by using `if / then` in the statements to blank out the zero terms; e.g. by changing the conclusion of `hasSum_mellin_pi_mul₀` from
````
HasSum (fun i ↦ π ^ (-s) * Gamma s * a i / p i ^ s) (mellin F s)
````
to
````
HasSum (fun i ↦ if i = 0 then 0 else π ^ (-s) * Gamma s * a i / p i ^ s) (mellin F s)
````
This also simplifies the proofs considerably. The approach I'm suggesting here is of course heavily based on your `LSeries.term`, but isn't quite covered by that, since `LSeries.term` only works for `Nat`s while I definitely want to allow `p i` to be an arbitrary non-negative real.

What do you think? I'd be happy with either option.

**Comment by MichaelStollBayreuth** [comment]
So far, the experience with how `LSeries.term` is set up was quite smooth. So the approach with blanking out the term at zero looks reasonable. Did you try to see if this leads to complications when you want to apply the results from this file?

**Comment by loefflerd** [comment]
> So far, the experience with how `LSeries.term` is set up was quite smooth. So the approach with blanking out the term at zero looks reasonable. Did you try to see if this leads to complications when you want to apply the results from this file?

I tried this setup, but unfortunately it turns out to be more troublesome than I expected. Developing a whole API for a version of `LSeries.term` with `NNReal` exponents seems overkill; and without that, every single proof ends up being weighted down with numerous `split_ifs`. I think I'd rather keep the code as it was before, with the `s ≠ 0` proofs inlined

**Comment by MichaelStollBayreuth** [comment]
OK, fine. Can you push your changes so that I can review them?

**Comment by loefflerd** [comment]
> OK, fine. Can you push your changes so that I can review them?

I ended up deciding I'd prefer the code as it currently stands.

**Comment by MichaelStollBayreuth** [comment]
I don't think you pushed the small stylistic changes (even though you remarked the conversations as "resolved" and at one point said that you had "golfed away" a part of the proof).

**Comment by loefflerd** [comment]
> I don't think you pushed the small stylistic changes (even though you remarked the conversations as "resolved" and at one point said that you had "golfed away" a part of the proof).

You're quite right, I pushed them to the wrong branch by mistake! Sorry.

**Review by MichaelStollBayreuth — APPROVED** [review]
LGTM!

**Comment by MichaelStollBayreuth** [comment]
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**